File: frontend/src/services/mapService.ts
/**
 * mapService — Data querying, filtering, and geolocation helpers for CarbonFlow Maps.
 * All business logic for filtering entities is centralized here, keeping components pure.
 * Future: replace mock imports with API calls to GET /api/v1/map/*
 */

import {
  SUPPLIERS,
  BUYERS,
  FACILITIES,
  ROUTES,
  REGIONS,
  CARBON_FLOW_EDGES,
} from '../data/mapsMockData';
import type {
  SupplierNode,
  BuyerNode,
  FacilityNode,
  RouteData,
  RegionData,
  CarbonFlowEdge,
  MapFilterState,
  GeoPoint,
} from '../types/maps';

// ─── Supplier Queries ─────────────────────────────────────────────────────────

export function getSuppliers(filters: MapFilterState, searchQuery: string): SupplierNode[] {
  return SUPPLIERS.filter((s) => {
    if (filters.region && !s.location.toLowerCase().includes(filters.region.toLowerCase())) return false;
    if (filters.industry && s.industry.toLowerCase() !== filters.industry.toLowerCase()) return false;
    if (s.purity < filters.minPurity) return false;
    if (s.availableTonnes < filters.minQuantity) return false;
    if (s.pricePerTon > filters.maxPrice) return false;
    if (filters.verifiedOnly && !s.verified) return false;
    if (searchQuery.trim()) {
      const q = searchQuery.toLowerCase();
      return (
        s.name.toLowerCase().includes(q) ||
        s.location.toLowerCase().includes(q) ||
        s.industry.toLowerCase().includes(q) ||
        s.facilityType.toLowerCase().includes(q)
      );
    }
    return true;
  });
}

// ─── Buyer Queries ────────────────────────────────────────────────────────────

export function getBuyers(filters: MapFilterState, searchQuery: string): BuyerNode[] {
  return BUYERS.filter((b) => {
    if (filters.region && !b.location.toLowerCase().includes(filters.region.toLowerCase())) return false;
    if (filters.application && b.application.toLowerCase() !== filters.application.toLowerCase()) return false;
    if (b.minPurity < filters.minPurity) return false;
    if (filters.verifiedOnly && !b.verified) return false;
    if (searchQuery.trim()) {
      const q = searchQuery.toLowerCase();
      return (
        b.name.toLowerCase().includes(q) ||
        b.organisation.toLowerCase().includes(q) ||
        b.location.toLowerCase().includes(q) ||
        b.application.toLowerCase().includes(q)
      );
    }
    return true;
  });
}

// ─── Facility Queries ─────────────────────────────────────────────────────────

export function getFacilities(_filters: MapFilterState, searchQuery: string): FacilityNode[] {
  return FACILITIES.filter((f) => {
    if (searchQuery.trim()) {
      const q = searchQuery.toLowerCase();
      return (
        f.name.toLowerCase().includes(q) ||
        f.location.toLowerCase().includes(q) ||
        f.operator.toLowerCase().includes(q)
      );
    }
    return true;
  });
}

// ─── Route Queries ────────────────────────────────────────────────────────────

export function getRoutes(filters: MapFilterState, searchQuery: string): RouteData[] {
  return ROUTES.filter((r) => {
    if (r.distanceKm > filters.maxDistance) return false;
    if (r.estimatedCostINR > filters.maxTransportCost) return false;
    if (searchQuery.trim()) {
      const q = searchQuery.toLowerCase();
      return (
        r.supplierName.toLowerCase().includes(q) ||
        r.buyerName.toLowerCase().includes(q) ||
        r.transportMode.toLowerCase().includes(q)
      );
    }
    return true;
  });
}

// ─── Region Queries ───────────────────────────────────────────────────────────

export function getRegions(filters: MapFilterState): RegionData[] {
  if (!filters.region) return REGIONS;
  return REGIONS.filter((r) => r.name.toLowerCase().includes(filters.region.toLowerCase()));
}

// ─── Carbon Flow Queries ──────────────────────────────────────────────────────

export function getCarbonFlows(): CarbonFlowEdge[] {
  return CARBON_FLOW_EDGES;
}

// ─── Search across all entities ───────────────────────────────────────────────

export interface SearchResult {
  id: string;
  label: string;
  subtitle: string;
  type: 'supplier' | 'buyer' | 'facility' | 'region';
  coords: GeoPoint;
}

export function searchEntities(query: string): SearchResult[] {
  if (!query.trim()) return [];
  const q = query.toLowerCase();
  const results: SearchResult[] = [];

  for (const s of SUPPLIERS) {
    if (s.name.toLowerCase().includes(q) || s.location.toLowerCase().includes(q)) {
      results.push({ id: s.id, label: s.name, subtitle: s.location, type: 'supplier', coords: s.coords });
    }
  }
  for (const b of BUYERS) {
    if (b.name.toLowerCase().includes(q) || b.location.toLowerCase().includes(q) || b.organisation.toLowerCase().includes(q)) {
      results.push({ id: b.id, label: b.name, subtitle: b.location, type: 'buyer', coords: b.coords });
    }
  }
  for (const f of FACILITIES) {
    if (f.name.toLowerCase().includes(q) || f.location.toLowerCase().includes(q)) {
      results.push({ id: f.id, label: f.name, subtitle: f.location, type: 'facility', coords: f.coords });
    }
  }
  for (const r of REGIONS) {
    if (r.name.toLowerCase().includes(q)) {
      results.push({ id: r.id, label: r.name, subtitle: `${r.supply.activeSuppliers} suppliers`, type: 'region', coords: r.coords });
    }
  }

  return results.slice(0, 10);
}

// ─── Geo utility ──────────────────────────────────────────────────────────────

/** Haversine distance between two GeoPoints in km */
export function distanceKm(a: GeoPoint, b: GeoPoint): number {
  const R = 6371;
  const dLat = ((b.lat - a.lat) * Math.PI) / 180;
  const dLng = ((b.lng - a.lng) * Math.PI) / 180;
  const s =
    Math.sin(dLat / 2) ** 2 +
    Math.cos((a.lat * Math.PI) / 180) * Math.cos((b.lat * Math.PI) / 180) * Math.sin(dLng / 2) ** 2;
  return R * 2 * Math.atan2(Math.sqrt(s), Math.sqrt(1 - s));
}

File: backend/app/services/map_service.py
from app.core.config import settings
from supabase import create_client, Client
from app.schemas.maps import SupplierNode, BuyerNode, FacilityNode, RouteData, CarbonFlowEdge, GeoPoint

class MapService:
    def __init__(self):
        self.db: Client = create_client(settings.SUPABASE_URL, settings.SUPABASE_SERVICE_ROLE_KEY)

    def get_suppliers(self):
        # We simulate suppliers using users with role=supplier and co2_listings
        res = self.db.table("co2_listings").select("*, users(*)").execute()
        suppliers = []
        for l in res.data:
            if l['users']:
                # mock coords for demo
                suppliers.append(SupplierNode(
                    id=l['id'],
                    coords=GeoPoint(lat=22.4707, lng=70.0577), # Jamnagar roughly
                    name=l['users']['company_name'] or "Unknown Supplier",
                    location="Jamnagar, GJ",
                    industry="Refining",
                    facilityType="Point-Source",
                    availableTonnes=l['volume_tpa'],
                    purity=l['purity_percentage'],
                    pricePerTon=l['price_per_ton'],
                    verified=True
                ))
        return suppliers

    def get_buyers(self):
        res = self.db.table("co2_requests").select("*, users(*)").execute()
        buyers = []
        for r in res.data:
            if r['users']:
                buyers.append(BuyerNode(
                    id=r['id'],
                    coords=GeoPoint(lat=17.3850, lng=78.4867), # Hyderabad
                    name=r['users']['company_name'] or "Unknown Buyer",
                    organisation="Deccan Cement",
                    location="Hyderabad, TS",
                    application=r['required_grade'],
                    minPurity=95.0,
                    verified=True
                ))
        return buyers

    def get_facilities(self):
        res = self.db.table("facilities").select("*").execute()
        facilities = []
        for f in res.data:
            facilities.append(FacilityNode(
                id=f['id'],
                coords=GeoPoint(lat=f['latitude'], lng=f['longitude']),
                name=f['name'],
                type=f['type'],
                operator="CarbonFlow Partner",
                location=f['region']
            ))
        return facilities

    def get_routes(self):
        res = self.db.table("routes").select("*").execute()
        routes = []
        for r in res.data:
            routes.append(RouteData(
                id=r['id'],
                supplierId=r['supplier_id'],
                buyerId=r['buyer_id'],
                supplierName=r['supplier_name'],
                buyerName=r['buyer_name'],
                geometry=[GeoPoint(lat=p['lat'], lng=p['lng']) for p in r['geometry']] if r['geometry'] else [],
                distanceKm=r['distance_km'] or 0,
                travelTimeHrs=r['travel_time_hrs'] or 0,
                estimatedCostINR=r['estimated_cost_inr'] or 0,
                transportMode=r['transport_mode'] or "Road",
                emissionsTco2e=r['emissions_tco2e'],
                reliabilityScore=r['reliability_score']
            ))
        return routes

    def get_carbon_flows(self):
        # We can map active orders to carbon flows
        res = self.db.table("orders").select("*").eq("status", "Continuous Flow").execute()
        flows = []
        for o in res.data:
            flows.append(CarbonFlowEdge(
                id=o['id'],
                fromId=o['supplier_id'],
                toId=o['buyer_id'],
                tonnes=o['volume'],
                active=True
            ))
        return flows
